ouroboros-consensus/src/Ouroboros/Consensus/BlockFetchServer.hs

```diff
@@ -57,30 +58,34 @@ blockFetchServer _tracer chainDB = senderSide
 
     receiveReq :: ChainRange blk
                -> m (BlockFetchBlockSender blk m ())
-    receiveReq chainRange = withIter chainRange $ \it ->
-      -- When we got an iterator, it will stream at least one block since its
-      -- bounds are inclusive (when the bounds are invalid, we won't get an
-      -- iterator and we reply with 'SendMsgNoBlocks'), so we don't have to
-      -- check whether the iterator is empty (and reply with
-      -- 'SendMsgNoBlocks').
-      return $ SendMsgStartBatch $ sendBlocks it
+    receiveReq chainRange = withIter chainRange $ \errIt ->
+      return $ case errIt of
+        -- The range is not in the ChainDB or it forks off more than @k@
+        -- blocks back.
+        Left  _  -> SendMsgNoBlocks $ return senderSide
+        -- When we got an iterator, it will stream at least one block since
+        -- its bounds are inclusive, so we don't have to check whether the
+        -- iterator is empty.
+        Right it -> SendMsgStartBatch $ sendBlocks it
 
 
     sendBlocks :: ChainDB.Iterator m blk
                -> m (BlockFetchSendBlocks blk m ())
     sendBlocks it = do
       next <- ChainDB.iteratorNext it
       case next of
-        IteratorExhausted      -> return $ SendMsgBatchDone (return senderSide)
+        IteratorExhausted      -> return $ SendMsgBatchDone $ return senderSide
         IteratorResult blk     -> return $ SendMsgBlock blk (sendBlocks it)
         IteratorBlockGCed hash -> throwM $ BlockGCed @blk hash
 
-    withIter :: ChainRange blk -> (ChainDB.Iterator m blk -> m a) -> m a
+    withIter :: ChainRange blk
+             -> (Either (UnknownRange blk) (ChainDB.Iterator m blk) -> m a)
+             -> m a
     withIter (ChainRange start end) = bracket
       -- TODO when streamBlocks throws an exception, we should return
       -- SendMsgNoBlocks. Which exception will the ChainDB throw?
```

Comment:
This TODO looks outdated now. Now that we report range errors up front, when an iterator throws an exception later then we can just let that propagate.